api/src/utils/apply-diff.ts

```diff
@@ -0,0 +1,305 @@
+import { Field, Relation, SchemaOverview } from '@directus/shared/types';
+import { Knex } from 'knex';
+import { CollectionsService, FieldsService, RelationsService } from '../services';
+import { ActionEventParams, Collection, MutationOptions, Snapshot, SnapshotDiff, SnapshotField } from '../types';
+import { getSchema } from './get-schema';
+import getDatabase from '../database';
+import { applyChange, Diff, DiffDeleted, DiffNew } from 'deep-diff';
+import { cloneDeep, merge, set } from 'lodash';
+import logger from '../logger';
+import emitter from '../emitter';
+
+type CollectionDelta = {
+	collection: string;
+	diff: Diff<Collection | undefined>[];
+};
+
+export async function applyDiff(
+	currentSnapshot: Snapshot,
+	snapshotDiff: SnapshotDiff,
+	options?: { database?: Knex; schema?: SchemaOverview }
+): Promise<void> {
+	const database = options?.database ?? getDatabase();
+	const schema = options?.schema ?? (await getSchema({ database, bypassCache: true }));
+
+	const nestedActionEvents: ActionEventParams[] = [];
+	const mutationOptions: MutationOptions = {
+		autoPurgeSystemCache: false,
+		bypassEmitAction: (params) => nestedActionEvents.push(params),
+	};
+
+	await database.transaction(async (trx) => {
+		const collectionsService = new CollectionsService({ knex: trx, schema });
+
+		const getNestedCollectionsToCreate = (currentLevelCollection: string) =>
+			snapshotDiff.collections.filter(
+				({ diff }) => (diff[0] as DiffNew<Collection>).rhs?.meta?.group === currentLevelCollection
+			) as CollectionDelta[];
+
+		const getNestedCollectionsToDelete = (currentLevelCollection: string) =>
+			snapshotDiff.collections.filter(
+				({ diff }) => (diff[0] as DiffDeleted<Collection>).lhs?.meta?.group === currentLevelCollection
+			) as CollectionDelta[];
+
+		const createCollections = async (collections: CollectionDelta[]) => {
+			for (const { collection, diff } of collections) {
+				if (diff?.[0].kind === 'N' && diff[0].rhs) {
+					// We'll nest the to-be-created fields in the same collection creation, to prevent
+					// creating a collection without a primary key
+					const fields = snapshotDiff.fields
+						.filter((fieldDiff) => fieldDiff.collection === collection)
```

Comment:
I knew this stuff looked familiar 😄 https://github.com/directus/directus/commit/ce104b6a9ccf7b9b1cccf820dbc3f5bbeafe1e94